src/dataflow/src/render/delta_join.rs

```diff
@@ -232,70 +262,80 @@ where
-
-                                    let (oks, errs) = build_filter(
-                                        update_stream,
-                                        &source_columns,
-                                        &mut predicates,
-                                        &mut equivalences,
-                                    );
-                                    update_stream = oks;
-                                    if let Some(errs) = errs {
-                                        region_errs.push(errs);
-                                    }
-
                                     bound_inputs.push(*other);
                                 }
 
-                                // We must now de-permute the results to return to the common order.
-                                // TODO: Non-demanded columns would need default values here.
-                                let permutation = (0 .. source_columns.len()).map(|c| {
-                                    source_columns.iter().position(|x| &c == x).expect("Did not find required column in output")
-                                }).collect::<Vec<_>>();
-                                update_stream = update_stream.map({
-                                    let mut row_packer = repr::RowPacker::new();
-                                    move |row| {
-                                        let datums = row.unpack();
-                                        row_packer.pack(permutation.iter().map(|c| datums[*c]))
-                                }});
+                                // We must now apply `mfp` as it *should* have sufficient support for
+                                // full evaluation. Before we do this, we need to permute it to refer
+                                // to the correct physical column locations.
+                                mfp.permute(&column_map, column_map.len());
+                                // At this point, `mfp` is very commonly the identity (we have fully
+                                // applied it in the previous closure). However, it may not be the
+                                // identity if it produces duplicate columns (i.e. its `projection`
+                                // member has duplicates). In this case, we aren't currently able to
+                                // optimize it away, as `columns` cannot represent some output column
+                                // being in two locations. There are many ways we could improve this,
+                                // including improving query optimization (we should rarely need to
+                                // produce two copies of the same column) and further improvements
+                                // here (figure out how to apply the projection in the prior closure).
+                                if !mfp.is_identity() {
+                                    // The `mfp` should just be projection at this point. If it is not,
```

Comment:
is that something we can assert on the mfp -- `is_only_projection` or something like that?